File: packages/shared/src/canonical.ts
/**
 * Canonical spec hashing — SPEC-022 / DR-034 (FR-3).
 *
 * One canonicalization contract, two twins. This is the Node twin; the Python
 * twin lives in `scripts/hooks/canonical.py` and MUST produce byte-identical
 * output (guarded by the INV-2 corpus-parity test). Keep the two in lockstep:
 * any change here must be mirrored there or INV-2 fails.
 *
 * The approved `specHash` covers a CANONICAL form of the spec, defined precisely
 * so independent implementations reproduce it byte-identically:
 *
 *   1. Normalize EOL to `\n` up front (deterministic before any splitting).
 *   2. Split the frontmatter block from the body.
 *   3. From the frontmatter, remove exactly the lifecycle keys `status` and
 *      `phases` (the latter including its indented child lines). Everything else
 *      — id, tier, type, epic, title, … — is content and is retained verbatim.
 *   4. Rejoin frontmatter-minus-lifecycle + body.
 *   5. Collapse ALL relative-link URLs to `](RELLINK)` (#252) — external,
 *      anchor and absolute links are kept; link text is kept.
 *   6. Normalize: strip trailing whitespace per line; collapse the trailing
 *      newline run to exactly one.
 *   7. sha256 the UTF-8 bytes; return the hex digest.
 *
 * Consequence (the fix): editing `status`/`phases` — the tool's own lifecycle
 * transitions — no longer voids a content approval. Editing the body or any
 * other frontmatter field still voids it (substantive change re-triggers review).
 *
 * Tier-0: depends only on `crypto`. No `vscode`, no `fs`, no network. This is a
 * pure string transform — it must NOT route through `parseSpec`, whose
 * default-injection and re-serialization would change bytes and break the
 * byte-for-byte contract.
 */

import * as crypto from 'crypto';

/**
 * Same anchor as `FRONTMATTER_RE` in `spec.ts` (and the Python twin's regex):
 * a leading `---\n`, lazily-captured block, then `\n---` with an optional
 * trailing newline. Operates on EOL-normalized (`\n`) input only.
 */
const FRONTMATTER_RE = /^---\n([\s\S]*?)\n---\n?/;

/** A top-level `status:` key line (column 0, no leading indent). */
const STATUS_LINE_RE = /^status[ \t]*:/;
/** A top-level `phases:` key line — opens the indented phases block. */
const PHASES_LINE_RE = /^phases[ \t]*:/;
/** Any indented line (a child of the open `phases:` block, or other nesting). */
const INDENTED_LINE_RE = /^[ \t]+/;

/**
 * Remove the lifecycle keys (`status`, `phases` + children) from a frontmatter
 * block, line-oriented. KEEPS every other line verbatim — including blank lines,
 * `#` comment lines, and every non-lifecycle key. The `phases:` block ends at the
 * first non-indented line (a new top-level key or a blank line), mirroring how the
 * YAML parser treats a nested block.
 *
 * A `status:`/`phases:` key with an inline comment (`status: implementing # x`) is
 * dropped wholesale — the whole line goes.
 */
function stripLifecycle(fm: string): string {
  const out: string[] = [];
  let inPhasesBlock = false;
  for (const line of fm.split('\n')) {
    if (inPhasesBlock) {
      // Still inside the phases block while the line is indented; a blank line or
      // a new top-level key ends it. (A blank line is NOT indented → ends block.)
      if (INDENTED_LINE_RE.test(line)) {
        continue; // drop child line
      }
      inPhasesBlock = false;
      // fall through to re-evaluate this line as a (possibly droppable) top-level key
    }
    if (PHASES_LINE_RE.test(line)) {
      inPhasesBlock = true;
      continue; // drop the `phases:` line itself
    }
    if (STATUS_LINE_RE.test(line)) {
      continue; // drop the `status:` line
    }
    out.push(line);
  }
  return out.join('\n');
}

/**
 * Compute the canonical string form of a raw spec. Pure, deterministic, and
 * byte-for-byte reproducible by the Python twin.
 */
export function canonicalizeSpec(raw: string): string {
  // 1. Normalize EOL up front.
  const normalized = raw.replace(/\r\n?/g, '\n');

  // 2. Split frontmatter from body.
  const m = normalized.match(FRONTMATTER_RE);
  let joined: string;
  if (!m) {
    // No frontmatter → body is the whole string.
    joined = normalized;
  } else {
    const fm = m[1];
    const body = normalized.slice(m[0].length);
    // 3. Remove lifecycle keys.
    const fmClean = stripLifecycle(fm);
    // 4. Rejoin.
    joined = '---\n' + fmClean + '\n---\n' + body;
  }

  // 5. Collapse relative-link URLs (#252). Automated directory renumbering
  //    (#83/#175) rewrites sibling-spec link *paths* without changing what is
  //    referenced; collapsing them stops that cry-wolf. ALL relative URLs are
  //    collapsed (`./…`, `../…`, bare-relative `child/x.md`, `file.md`); external
  //    (`scheme://…`, `mailto:`), anchors (`#…`) and absolute (`/…`) are KEPT.
  //    Link *text* is preserved, so changing the referenced spec is still caught.
  //    MUST stay byte-identical to the Python twin's regex.
  const linked = joined.replace(/\]\((?![a-z][a-z0-9+.-]*:)(?!#)(?!\/)[^)]*\)/gi, '](RELLINK)');

  // 6. Normalize per-line trailing whitespace (the design's `trimEnd()` — strips
  //    all trailing whitespace; lines are already newline-free after the split, so
  //    this matches Python's `str.rstrip()` byte-for-byte), then collapse the
  //    trailing newline run to exactly one.
  const lines = linked.split('\n').map((ln) => ln.replace(/\s+$/, ''));
  return lines.join('\n').replace(/\n*$/, '') + '\n';
}

/** sha256 hex digest of the canonical form (FR-3). */
export function specHash(raw: string): string {
  return crypto.createHash('sha256').update(Buffer.from(canonicalizeSpec(raw), 'utf-8')).digest('hex');
}

// Two boundaries: canonical = frontmatter-minus-lifecycle + body (specHash); getSpecBodyOnly = body only.
/**
 * SPEC-017 FR-4 baseline boundary: the spec body AFTER the frontmatter block.
 *
 * Reuses the SAME `FRONTMATTER_RE` anchor `canonicalizeSpec` (and `parseSpec`)
 * use, so there is exactly ONE body-split definition — no second anchor to drift.
 * EOL is normalized to `\n` first (matching `parseSpec`). A spec with no
 * frontmatter returns the whole (normalized) content as body.
 *
 * Distinct from the canonical-hash boundary on purpose: `specHash` keeps
 * frontmatter-minus-lifecycle so editing `status`/`phases` does not void approval;
 * `getSpecBodyOnly` excludes ALL frontmatter so frontmatter churn never registers
 * as the human reworking the LLM's prose. Pure, Tier-0 (no new dependency).
 */
export function getSpecBodyOnly(raw: string): string {
  const normalized = raw.replace(/\r\n?/g, '\n');
  const m = normalized.match(FRONTMATTER_RE);
  return m ? normalized.slice(m[0].length) : normalized;
}

File: scripts/hooks/canonical.py
#!/usr/bin/env python3
"""Canonical spec hashing — SPEC-022 / DR-034 (FR-3). Python twin.

Byte-for-byte twin of `packages/shared/src/canonical.ts`. The PreToolUse gate
(`spec-gate.py`) is Python and cannot import the Node package, so it imports this
module instead. INV-2 (the corpus-parity test) asserts this twin and the Node
module produce byte-identical output over every spec in `specs/`.

Keep the two in lockstep: any change to the Node module must be mirrored here, or
INV-2 fails. The algorithm (see the Node module for the full prose):

  1. Normalize EOL to \n up front.
  2. Split the frontmatter block from the body.
  3. Remove exactly the lifecycle keys `status` and `phases` (+ indented children).
  4. Rejoin frontmatter-minus-lifecycle + body.
  5. Strip per-line trailing whitespace; collapse the trailing newline run to one.
  6. sha256 the UTF-8 bytes; return the hex digest.

Pure stdlib (`re`, `hashlib`). Also runnable as a CLI: `canonical.py --hash <file>`
prints the hex digest of the file's canonical form (used by the parity test).
"""
import hashlib
import re
import sys

# Same anchor as the Node FRONTMATTER_RE and spec.ts: leading `---\n`, lazy block,
# then `\n---` with an optional trailing newline. Operates on \n-normalized input.
_FRONTMATTER_RE = re.compile(r'^---\n([\s\S]*?)\n---\n?')

_STATUS_LINE_RE = re.compile(r'^status[ \t]*:')
_PHASES_LINE_RE = re.compile(r'^phases[ \t]*:')
_INDENTED_LINE_RE = re.compile(r'^[ \t]+')


def _strip_lifecycle(fm: str) -> str:
    """Drop `status` line + `phases` block (its indented children), KEEP all else."""
    out = []
    in_phases_block = False
    for line in fm.split('\n'):
        if in_phases_block:
            if _INDENTED_LINE_RE.match(line):
                continue  # drop child line
            in_phases_block = False
            # fall through: re-evaluate this line as a top-level key
        if _PHASES_LINE_RE.match(line):
            in_phases_block = True
            continue  # drop the `phases:` line itself
        if _STATUS_LINE_RE.match(line):
            continue  # drop the `status:` line
        out.append(line)
    return '\n'.join(out)


def canonicalize_spec(raw: str) -> str:
    """Compute the canonical string form of a raw spec (mirrors canonicalizeSpec)."""
    # 1. Normalize EOL up front.
    normalized = re.sub(r'\r\n?', '\n', raw)

    # 2. Split frontmatter from body.
    m = _FRONTMATTER_RE.match(normalized)
    if not m:
        joined = normalized
    else:
        fm = m.group(1)
        body = normalized[m.end():]
        # 3. Remove lifecycle keys. 4. Rejoin.
        fm_clean = _strip_lifecycle(fm)
        joined = '---\n' + fm_clean + '\n---\n' + body

    # 5. Strip per-line trailing whitespace (str.rstrip() == JS trimEnd over a
    #    newline-free line), then collapse the trailing newline run to exactly one.
    lines = [ln.rstrip() for ln in joined.split('\n')]
    return re.sub(r'\n*$', '', '\n'.join(lines)) + '\n'


def spec_hash(raw: str) -> str:
    """sha256 hex digest of the canonical form (FR-3)."""
    return hashlib.sha256(canonicalize_spec(raw).encode('utf-8')).hexdigest()


def _main(argv):
    if len(argv) == 3 and argv[1] == '--hash':
        with open(argv[2], 'r', encoding='utf-8') as fh:
            sys.stdout.write(spec_hash(fh.read()))
        return 0
    sys.stderr.write('usage: canonical.py --hash <file>\n')
    return 2


if __name__ == '__main__':
    sys.exit(_main(sys.argv))
